File: Leetcode1292/Solution.cpp
// optimized solution with enumeration

class Solution {
public:
    int getRect(const vector<vector<int>>& P, int x1, int y1, int x2, int y2) {
        return P[x2][y2] - P[x1 - 1][y2] - P[x2][y1 - 1] + P[x1 - 1][y1 - 1];
    }

    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        int m = mat.size(), n = mat[0].size();
        vector<vector<int>> P(m + 1, vector<int>(n + 1));
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                P[i][j] = P[i - 1][j] + P[i][j - 1] - P[i - 1][j - 1] +
                          mat[i - 1][j - 1];
            }
        }

        int r = min(m, n), ans = 0;
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                for (int c = ans + 1; c <= r; ++c) {
                    if (i + c - 1 <= m && j + c - 1 <= n &&
                        getRect(P, i, j, i + c - 1, j + c - 1) <= threshold) {
                        ++ans;
                    } else {
                        break;
                    }
                }
            }
        }
        return ans;
    }
};

// normal solution using prefix sum and binary search
class Solution {
public:
    int getRect(const vector<vector<int>>& P, int x1, int y1, int x2, int y2) {
        return P[x2][y2] - P[x1 - 1][y2] - P[x2][y1 - 1] + P[x1 - 1][y1 - 1];
    }

    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        int m = mat.size(), n = mat[0].size();
        vector<vector<int>> P(m + 1, vector<int>(n + 1));
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                P[i][j] = P[i - 1][j] + P[i][j - 1] - P[i - 1][j - 1] +
                          mat[i - 1][j - 1];
            }
        }

        int l = 1, r = min(m, n), ans = 0;
        while (l <= r) {
            int mid = (l + r) / 2;
            bool find = false;
            for (int i = 1; i <= m - mid + 1; ++i) {
                for (int j = 1; j <= n - mid + 1; ++j) {
                    if (getRect(P, i, j, i + mid - 1, j + mid - 1) <=
                        threshold) {
                        find = true;
                    }
                }
            }
            if (find) {
                ans = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return ans;
    }
};

// runtime - 1ms
class Solution {
public:

    void debug(const vector<vector<int>>& mat)
    {
        for (const vector<int>& row: mat)
        {
            for (const int& col: row)
            {
                cout << col << " ";
            }
            cout << endl;         
        }
    }

    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        vector<vector<int>> prefix_sum(mat.size() + 1, vector<int>(mat[0].size() + 1, 0));

        for (int ii = 1; ii <= mat.size(); ++ii)
        {
            for (int jj = 1; jj <= mat[0].size(); ++jj)
            {
                prefix_sum[ii][jj] = mat[ii-1][jj-1] +  prefix_sum[ii-1][jj] + prefix_sum[ii][jj - 1] -  prefix_sum[ii-1][jj-1];
            }
        }
        
        int max_side = 0;
        for (int ii = 1; ii <= mat.size(); ++ii)
        {
            for (int jj = 1; jj <= mat[0].size(); ++jj)
            {
                if (max_side < ii && max_side < jj)
                {
                    int k = max_side + 1;
                    int curr_sum = prefix_sum[ii][jj] + prefix_sum[ii-k][jj-k] - prefix_sum[ii-k][jj] - prefix_sum[ii][jj-k];
                    if (curr_sum <= threshold)
                    {
                        max_side = k; 
                    } 
                }
            }    
        }

        return max_side;
    }
};


// runtime - 5ms
class Solution {
public:
    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        int rows = mat.size();
        int cols = mat[0].size();

        vector<vector<int>> prefix(rows, vector<int>(cols, 0));

        // Build prefix sum
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                prefix[i][j] = mat[i][j]
                             + (i > 0 ? prefix[i - 1][j] : 0)
                             + (j > 0 ? prefix[i][j - 1] : 0)
                             - (i > 0 && j > 0 ? prefix[i - 1][j - 1] : 0);
            }
        }

        // Helper lambda to get sum of square
        auto sumSquare = [&](int i, int j, int r2, int c2) {
            int sum = prefix[r2][c2];
                    if (i > 0) sum -= prefix[i - 1][c2];
                    if (j > 0) sum -= prefix[r2][j - 1];
                    if (i > 0 && j > 0) sum += prefix[i - 1][j - 1];
            
            return sum;
        };

        int best = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                for (int k = best; k < min(rows - i, cols - j); k++) { //offset to find bottom right cell
                    //bottom right cell (r2, c2)
                    int r2 = i + k;
                    int c2 = j + k;

                    int sum = sumSquare(i, j, r2, c2);

                    if (sum <= threshold) {
                        best = k + 1; //(offset + 1) gives the side of square
                    } else {
                        break; //because sum will increase only. Better move to next cell
                    }
                }
            }
        }

        return best;
    }
};

File: Leetcode1292/Solution.py
# optimized solution with optimization and enumeration
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        P = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                P[i][j] = (
                    P[i - 1][j]
                    + P[i][j - 1]
                    - P[i - 1][j - 1]
                    + mat[i - 1][j - 1]
                )

        def getRect(x1, y1, x2, y2):
            return P[x2][y2] - P[x1 - 1][y2] - P[x2][y1 - 1] + P[x1 - 1][y1 - 1]

        r, ans = min(m, n), 0
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                for c in range(ans + 1, r + 1):
                    if (
                        i + c - 1 <= m
                        and j + c - 1 <= n
                        and getRect(i, j, i + c - 1, j + c - 1) <= threshold
                    ):
                        ans += 1
                    else:
                        break
        return ans

class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        P = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                P[i][j] = (
                    P[i - 1][j]
                    + P[i][j - 1]
                    - P[i - 1][j - 1]
                    + mat[i - 1][j - 1]
                )

        def getRect(x1, y1, x2, y2):
            return P[x2][y2] - P[x1 - 1][y2] - P[x2][y1 - 1] + P[x1 - 1][y1 - 1]

        l, r, ans = 1, min(m, n), 0
        while l <= r:
            mid = (l + r) // 2
            find = any(
                getRect(i, j, i + mid - 1, j + mid - 1) <= threshold
                for i in range(1, m - mid + 2)
                for j in range(1, n - mid + 2)
            )
            if find:
                ans = mid
                l = mid + 1
            else:
                r = mid - 1
        return ans

class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        side = 0
        
        # prefix sum matrix
        prefix_sum = [[0]*(n+1) for _ in range(m+1)]

        # build the prefix sums
        for i in range(1, m+1):
            for j in range(1, n+1):
                prefix_sum[i][j] = (mat[i-1][j-1] + prefix_sum[i-1][j] + prefix_sum[i][j-1] - prefix_sum[i-1][j-1])

        # maximum possible square size
        max_side = min(m, n)

        # try square from largest to smallest
        while max_side > 0:
            for start_row in range(m-max_side + 1):
                for start_col in range(n - max_side + 1):
                    square_sum = (
                        prefix_sum[start_row + max_side][start_col + max_side]
                        - prefix_sum[start_row][start_col + max_side]
                        - prefix_sum[start_row + max_side][start_col]
                        + prefix_sum[start_row][start_col]
                    )
                    if(square_sum <= threshold):
                        return max_side
            max_side -= 1

        return 0


# runtime - 102ms
from typing import List

class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        # build prefix sum matrix
        pre = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
            row_sum = 0
            for j in range(n):
                row_sum += mat[i][j]
                pre[i + 1][j + 1] = pre[i][j + 1] + row_sum

        def square_exists(k: int) -> bool:
            # check if any k x k square has sum <= threshold
            for i in range(k, m + 1):
                for j in range(k, n + 1):
                    s = (
                        pre[i][j]
                        - pre[i - k][j]
                        - pre[i][j - k]
                        + pre[i - k][j - k]
                    )
                    if s <= threshold:
                        return True
            return False

        ans = 0
        # grow side length while feasible
        for k in range(1, min(m, n) + 1):
            if square_exists(k):
                ans = k
            else:
                break
        return ans


# runtime - 158ms
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        prefix = [[0] * (len(mat[0])+1)]
        for row in mat:
            prefix.append([0]+row)
            for idx in range(len(row)):
                prefix[-1][idx+1] -=prefix[-2][idx] 
                prefix[-1][idx+1] +=prefix[-1][idx] 
                prefix[-1][idx+1] +=prefix[-2][idx+1] 

        result = 0
        for r in range(len(prefix)-1):
            for c in range(len(prefix[0])-1):
                end_r = r + result+1
                end_c = c+ result+1
                while end_r < len(prefix) and end_c < len(prefix[0]) and prefix[end_r][end_c] - prefix[r][end_c] - prefix[end_r][c] + prefix[r][c] <=threshold:
                    end_r+=1
                    end_c+=1
                    result +=1
        return result  

# runtime - 214ms
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])

        prefix = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m):
            for j in range(n):
                prefix[i + 1][j + 1] = (
                    prefix[i][j + 1]
                    + prefix[i + 1][j]
                    - prefix[i][j]
                    + mat[i][j]
                )

        left, right = 1, min(m, n)
        ans = 0

        while left <= right:
            mid = (left + right) // 2
            isValid = False

            for i in range(mid, m + 1):
                for j in range(mid, n + 1):
                    total = (
                        prefix[i][j]
                        - prefix[i - mid][j]
                        - prefix[i][j - mid]
                        + prefix[i - mid][j - mid]
                    )
                    if total <= threshold:
                        isValid = True
                        break
                if isValid:
                    break

            if isValid:
                ans = mid
                left = mid + 1
            else:
                right = mid - 1

        return ans

# runtime - 270ms
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])

        # =========================
        # Step 1: 构建二维前缀和
        # =========================
        prefix = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m):
            for j in range(n):
                prefix[i + 1][j + 1] = (
                    prefix[i][j + 1]
                    + prefix[i + 1][j]
                    - prefix[i][j]
                    + mat[i][j]
                )

        # 计算以 (i, j) 为左上角，边长为 k 的正方形和
        def square_sum(i: int, j: int, k: int) -> int:
            return (
                prefix[i + k][j + k]
                - prefix[i][j + k]
                - prefix[i + k][j]
                + prefix[i][j]
            )

        # =========================
        # Step 2: 二分正方形边长
        # =========================
        left, right = 0, min(m, n)

        while left < right:
            mid = (left + right + 1) // 2
            found = False

            for i in range(m - mid + 1):
                for j in range(n - mid + 1):
                    if square_sum(i, j, mid) <= threshold:
                        found = True
                        break
                if found:
                    break

            if found:
                left = mid
            else:
                right = mid - 1

        return left



# runtime - 326ms
class Solution:
    def check(self, mat, bound, threshold, rows, cols):
        startRow, startCol = bound, bound
        for i in range(startRow, rows):
            for j in range(startCol, cols):
                val = mat[i][j]
                if i - bound - 1 >= 0:
                    val -= mat[i-bound-1][j]
                if j - bound - 1 >= 0:
                    val -= mat[i][j-bound-1]
                if i - bound - 1 >= 0 and j - bound - 1 >= 0:
                    val += mat[i-bound-1][j-bound-1]
                
                if val <= threshold:
                    return True

        return False

    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        rows, cols = len(mat), len(mat[0])
        for i in range(1, cols):
            mat[0][i] += mat[0][i-1]
        for i in range(1, rows):
            mat[i][0] += mat[i-1][0]
        for i in range(1, rows):
            for j in range(1, cols):
                mat[i][j] += (mat[i-1][j] + mat[i][j-1] - mat[i-1][j-1])
    
        left, right = 0, max(rows, cols)+1
        while left < right:
            mid = (left + right) // 2
            maxLength = min(rows, cols)
            if self.check(mat, mid, threshold, rows, cols):
                left = mid + 1
            else:
                right = mid
        
        return left
